Realm/RLMSchema.mm

```diff
@@ -71,66 +91,92 @@ + (void)initialize {
     }
     initialized = true;
 
-    NSMutableArray *schemaArray = [NSMutableArray array];
-    RLMSchema *schema = [[RLMSchema alloc] init];
+    s_localNameToClass = [NSMutableDictionary dictionary];
+    s_partialSharedSchema = [[RLMSchema alloc] init];
+}
 
-    unsigned int numClasses;
-    Class *classes = objc_copyClassList(&numClasses);
++ (instancetype)partialSharedSchema {
+    return s_partialSharedSchema;
+}
 
-    // first create class to name mapping so we can do array validation
-    // when creating object schema
-    s_localNameToClass = [NSMutableDictionary dictionary];
-    for (unsigned int i = 0; i < numClasses; i++) {
-        Class cls = classes[i];
-        static Class objectBaseClass = [RLMObjectBase class];
-        if (!RLMIsKindOfClass(cls, objectBaseClass) || ![cls shouldPersistToRealm]) {
-            continue;
++ (void)registerClasses:(Class *)classes count:(NSUInteger)count {
+    @synchronized(s_localNameToClass) {
+        // first create class to name mapping so we can do array validation
+        // when creating object schema
+        for (NSUInteger i = 0; i < count; i++) {
+            Class cls = classes[i];
+
+            if (!RLMIsObjectSubclass(cls)) {
+                continue;
+            }
+
+            NSString *className = NSStringFromClass(cls);
+            if ([RLMSwiftSupport isSwiftClassName:className]) {
+                className = [RLMSwiftSupport demangleClassName:className];
+            }
+            // NSStringFromClass demangles the names for top-level Swift classes
+            // but not for nested classes. _T indicates it's a Swift symbol, t
+            // indicates it's a type, and C indicates it's a class.
+            else if ([className hasPrefix:@"_TtC"]) {
+                NSString *message = [NSString stringWithFormat:@"RLMObject subclasses cannot be nested within other declarations. Please move %@ to global scope.", className];
+                @throw RLMException(message);
+            }
+
+            if (Class existingClass = s_localNameToClass[className]) {
+                if (existingClass != cls) {
+                    NSString *message = [NSString stringWithFormat:@"RLMObject subclasses with the same name cannot be included twice in the same target. Please make sure '%@' is only linked once to your current target.", className];
+                    @throw RLMException(message);
+                }
+                continue;
+            }
+
+            s_localNameToClass[className] = cls;
+
+            RLMReplaceClassNameMethod(cls, className);
+            // override sharedSchema classs method to return nil to avoid topo-sort issues
+            RLMReplaceSharedSchemaMethod(cls, nil);
```

Comment:
No, because calling sharedSchema will cause us to hit this lock if the shared schema isn't loaded yet